tempesta_fw/classifier/frang.c

```diff
@@ -0,0 +1,695 @@
+		}
+	}
+
+	return TFW_PASS;
+}
+
+static int
+frang_http_body_len_limit(const TfwHttpReq *req)
+{
+	if (frang_cfg.http_body_len &&
+	    tfw_str_len(&req->body) > frang_cfg.http_body_len) {
+		TFW_DBG("frang: http_body_len limit is reached\n");
+		return TFW_BLOCK;
+	}
+
+	return TFW_PASS;
+}
+
+static int
+frang_http_len_limit(const TfwHttpReq *req)
+{
+	int r;
+
+	r = frang_http_uri_len_limit(req);
+	if (r)
+		return r;
+	r = frang_http_field_len_limit(req);
+	if (r)
+		return r;
+	r = frang_http_body_len_limit(req);
+	if (r)
+		return r;
+
+	return 0;
+}
+
+static int
+frang_http_methods_check(const TfwHttpReq *req)
+{
+	unsigned long m = (1 << req->method);
+
+	if (frang_cfg.http_methods_mask && (frang_cfg.http_methods_mask & m))
+		return TFW_PASS;
+
+	TFW_DBG("frang: forbidden method: %d (%#lx)\n", req->method, m);
+	return TFW_BLOCK;
+}
+
+static int
+frang_http_ct_check(const TfwHttpReq *req)
+{
+#define _CT "Content-Type"
+#define _CTLEN (sizeof(_CT) - 1)
+	TfwStr *ct, *end;
+	FrangCtVal *curr;
+
+	if (!frang_cfg.http_ct_is_required || !frang_cfg.http_ct_vals ||
+	    req->method != TFW_HTTP_METH_POST)
+		return TFW_PASS;
+
+	/* Find the Content-Type header.
+	 *
+	 * XXX: Should we make the header "special"?
+	 * It would speed up this function, but bloat the HTTP parser code,
+	 * and pollute the headers table.
+	 */
+	TFW_HTTP_FOR_EACH_RAW_HDR_FIELD(ct, end, req) {
+		if (tfw_str_eq_cstr(ct, _CT, _CTLEN, TFW_STR_EQ_PREFIX_CASEI))
+			break;
+
+	}
+	if (ct == end) {
+		TFW_DBG("frang: Content-Type is missing\n");
+		return TFW_BLOCK;
+	}
+
+	/* Check that the Content-Type is in the list of allowed values.
+	 *
+	 * TODO: possible improvement: binary search.
+	 * Generally the binary search is more efficient, but linear search is
+	 * usually faster for small sets of values. Perhaps we should switch
+	 * between two if the performance is that critical here, but benchmarks
+	 * should be done to measure the impact.
+	 *
+	 * TODO: don't store field name in the TfwStr. Store only the header
+	 * value, and thus get rid of the nasty tfw_str_eq_kv().
+	 */
+	for (curr = frang_cfg.http_ct_vals; curr->str; ++curr) {
+		if (tfw_str_eq_kv(ct, _CT, _CTLEN, ':', curr->str, curr->len,
+				  TFW_STR_EQ_PREFIX_CASEI))
+			break;
+	}
+
+	if (!curr->str) {
+		TFW_DBG("frang: forbidden Content-Type value\n");
+		return TFW_BLOCK;
+	}
+
+	return TFW_PASS;
+#undef _CT
+#undef _CTLEN
+}
+
+static int
+frang_http_host_check(const TfwHttpReq *req)
+{
+	TfwStr *field;
+
+	if (!frang_cfg.http_host_is_required || req->method != TFW_HTTP_METH_POST)
+		return TFW_PASS;
+
+	field = &req->h_tbl->tbl[TFW_HTTP_HDR_HOST].field;
+	if (!field->ptr) {
+		TFW_DBG("frang: the Host header is missing\n");
+		return TFW_BLOCK;
+	}
+
+	/* FIXME: here should be a check that the Host value is not an IP
+	 * address. Need a fast routine that supports compound TfwStr.
+	 * Perhaps should implement a tiny FSM or postpone the task until we
+	 * have a good regex library. */
```

Comment:
Propose a quick solution. Probably we'll be Ok to live with it. I believe nothing special from TfwStr isn't required and you can do this with current implementation.